design/034-new_lockfiles.md

```diff
@@ -0,0 +1,237 @@
+# Proposal: New Conan 2.0 lockfiles
+
+| **Status**        |                                                   |
+|:------------------|:--------------------------------------------------|
+| **RFC #**         | [034](https://github.com/conan-io/tribe/pull/34)  |
+| **Submitted**     | 2022-04-26                                        |
+| **Tribe votes**   |                                                   |
+
+
+## Summary
+
+Conan 2.0 will implement completely new lockfiles, with a new proposal of format, definition, behavior, usage and flow, based on the following principles:
+
+- Lockfiles files will be json files containing only three lists of ordered references. One list for “host” requires, another list for “build” requires, another list for “python” requires.
+- Lockfiles will no longer contain information about profiles, settings, options (as they did in 1.X)
+- Lockfiles will not contain information about the dependency graph, only ordered lists (ordered by version and revision timestamp) of package references.
+- The default level of locking will be locking down to the recipe reference, that is, including the version and the recipe revision (``pkg/version@user/channel#recipe_revision``), but not the package-id nor the package-revision. This is aligned with the previously accepted Tribe proposal of removing the [``package_revision_mode``](https://github.com/conan-io/tribe/pull/30). Even if implementing lockfiles locking down the package revision will be possible, that will be considered the exception, and the main flows, documentation and behavior will be optimized for locking down to the recipe revision.
+- Locking by default will be non-strict, that is, if some ``requires`` cannot find a matching locked version in the lockfile, it will be resolved normally. A ``--lockfile-strict`` mode will be implemented, but not the default, to enforce finding a match for declared requires in the lockfile or failing otherwise.
```

Comment:
Strict enforcement of the lock is most likely the single most reported issue in our support, and the origin of many frustrations. Recall that many operations imply in many occasions that it cannot be strict:
- Capturing a lockfile for every configuration after the first one, when the first lockfile from the first configuration is provided, to guarantee it is as consistent as possible between configurations. So if we have to capture a lockfile for N configurations, that means that N-1 ``conan lock create`` commands will have to add ``--lockfile-no-strict``
- Every change that every developer does to a package, can potentially change the dependencies. Add a new one, update the version of a existing one, etc. This is a very common operation in the day-to-day. This also requires ``--lockfile-no-strict`` otherwise each developer change to ``requires`` will raise an error.
- Application of a lockfile created with ``conan lock add`` to lock only some of the dependencies or transitive dependencies will also require ``--lockfile-no-strict``, always, by definition.
- The application of a partial lockfile, like the one used to build the changes done to a package in the middle of the graph, when we want to apply it to a downstream consumer application, will also require ``--lockfile-no-strict``, always, by definition

One of the first implementation we had was the other way around, strict by default. When we realized that we were basically either saying ``strict=False`` in code (for some commands), or passing ``--lockfile-no-strict`` in the command line for the majority of command lines we had in our testing and examples (note that we have been working in more elaborte examples like https://github.com/conan-io/examples2/pull/7, to further learn and evaluate this proposal), we decided to change the default.

Note that being non-strict doesn't mean that it will not lock things correctly. Whatever was locked will be correctly locked, unless the developer explicitly wanted to change dependencies outside of that lock. Lockfiles will prevent against any changes in the dependencies, if new versions or revisions are published. They never intended to be (and they are not in other language package managers) a mechanism to block developers to do modifications to source and recipes, and this was probably the root source of the above mentioned frustrations.

Reply:
My thoughts exactly. My intuition tells me that lockfiles should be honored 100% of the times to ensure reproducible builds.